CRC feature implementation

deepikabhavnani opened the pull request:
CRC class implementation

CRC class `BaseCRC.h` is created to support hardware/software CRCs

BitwiseCRC.h contains software CRC implementation for bitwise CRC computation which will be somewhat slow compared to other techniques as bit by bit computation is performed in software.

FastCRC.h contains software CRC implementation using lookup table method. Table is generated during init step.

HalfByteCRC.h constains software implementation for CRC32 using 16-entry technique to have limited table size (64bytes) and is fast compared to bitwise CRC computation. Table is generated runtime during init call.

Usage example:
https://gist.github.com/deepikabhavnani/0dacf5572c0b4d87b8fbfbcfd9246a95

geky commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,102 @@
+
+#define GET_DATA_SIZE(x) ((x) <= 8 ? 1 : ((x) <= 16 ? 2 : 4))
+#define TOP_BIT(x)       ((x) < 8 ? (1u << 7) : (1u << (x - 1)))
+#define DATA_MASK(x)     ((1u << x) - 1)
```
Can these defines be moved into the C++ file?

deepikabhavnani replied:
Yes

0xc0170 replied:
If we keep this as defines (I would do local inline functions helpers to get this would have the same result and might not pollute symbols), then use prefix as for above enums (`CRC_`).

hanno-arm replied:
There are two uses of `x` that are not enclosed in parentheses - could you add those?

geky commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,202 @@
+            case CRC_7BIT_SD:
+                _polynomial = 0x9;             // x7+x3+1;
+                _width = 7;
+                break;
```
Why not just provide polynomial/width/initial_value/reflect_data/reflect_remainder/final_xor all as constructor parameters with reasonable defaults?

If you want to get creative, you could do something like this:
``` cpp
// Presets for standard 32 bit CRC
#define CRC_32BIT 0x04C11DB7, 32, ~0, true, true, ~0

// Declare 32 bit CRC
SlowCRC<uint32_t> crc(CRC_32BIT);

// Declare non-standard 32 bit CRC
SlowCRC<uint32_t> crc(0x04C11DB7, 32, 0);
```

deepikabhavnani replied:
Because not all of them are verified, I don't want to give impression that all CRC computations are implemented without verifying each of them. Ones which are used most are implemented and verified. Deviation from standard ones will give incorrect result.

geky replied:
> Deviation from standard ones will give incorrect result.

They're all backed by the same algorithm though aren't they?

You could still say only the CRCs with matching defines are supported, and assert for invalid configurations in the constructor.

geky commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,102 @@
+/** @{*/
+#include <stdint.h>
+
+typedef uint64_t data_size_t;
```
Should this be crc_data_size_t? or just crc_t?

data_size_t is ambiguous

deepikabhavnani replied:
crc_data_size_t 👍 . Will change it

geky commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,102 @@
+     *
+     *  @param  buffer  Data bytes     
+     *  @param  size  Size of data
+     *  @param  crc  CRC 
```
Thought: you should document what the initial state of crc is expected to be.

Can I crc data without storing it all in memory at once? This is a requirement for littlefs.

deepikabhavnani replied:
Initial state of CRC is not picked from this argument, we only return computed CRC , Initial value is don;t care. 
Inital XOR values for CRC, vary based on the CRC algorithm selected, which is set as part of constructor.

geky replied:
Do you think we could have a `compute_partial` or something similar to continue a CRC without initializing the value?

deepikabhavnani replied:
Partial CRC, yes we can have support for that in future. But partial CRC computation should be separate API and we will need proper locking mechanism with that. Partial CRC method is used to compute entire CRC in parts in parallel.

geky replied:
Oh no, not for parallel computation. Just for computing a CRC when the data being CRC'd doesn't fit in RAM. The data can be streamed in, but only if the CRC can be computed partially.

This can be another pr if you'd like, it would just be needed before this class can be adopted everywhere.

geky commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,202 @@
+#include "mbed.h"
+
+template <typename T>
+class FastCRC : public CRC
```
It may be worth documenting this template parameter. I'm not entirely sure what value it adds to the class (the internal table uses T as elements? Does the resulting CRC size depend on T or the polynomial argument?)

deepikabhavnani replied:
Resulting CRC size will always be uint32_t  (masked based on CRC width). Kept it generic in CRC class

geky replied:
Does this need to be a template then? you can extend an instantiated template:
``` cpp
class FastCRC : public CRC<uint32_t>
```

deepikabhavnani replied:
CRC class is generic and will be used for hardware CRC as well. If you check hardware CRC implementation of target devices, they all return 32-bit CRC (masked) based on polynomial. Could be because of C implementation, but in order to have consistent behavior it would be good to not have template in base class.

Any inputs from HAL team?

geky commented:
Is there a CRC implementation that sits between FastCRC and SlowCRC in terms of ROM vs runtime?

I'm curious which CRC option best matches littlefs's:
https://github.com/ARMmbed/mbed-os/blob/master/features/filesystem/littlefs/littlefs/lfs_util.c#L21-L35

Also is it safe to assume this opens up the opportunity to add hardware accelerated CRC?

deepikabhavnani commented:
Intermediate version is to use smaller tables (16-entry table) if we have space constraints. The method which you have in littlefs.

0xc0170 commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,102 @@
+/* mbed Microcontroller Library
```
location - `features/crc` ? this is not part of platform folder? - just thinking why feature/ location was chosen

deepikabhavnani replied:
Current CRC use case is block device and filesystem, hence selected feature. 
In future mbed_crc.h class can be used as base class for hardware CRC, hence we can have it HAL layer as well. 
Inputs from core-team will be appreciated for this.

bulislaw replied:
I would assume it should go to drivers/ straight away.

deepikabhavnani replied:
move entire CRC implementation to drivers? Or just mbed_crc.h.
Naming convention in driver folder does not have `mbed` prefix. Should I rename file to be CRC.h?
If everything is moved inside driver, should we have folder structure for software implementation or everything should be at same level?

bulislaw replied:
I would say all CRC files and we should match the convention for the drivers and I don't think we need directory structure for it.

0xc0170 commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,202 @@
+#define MBED_FAST_CRC_H
+
+#include "mbed_crc.h"
+#include "mbed.h"
```
This should not be included within our code base (include everything is not an answer to what symbol is missing), please include what is needed here.

geky replied:
Good catch 👍 

Sounds like we should have a script to catch these... 🤔

0xc0170 commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,202 @@
+     *
+     * @return  0 on success, negative error code on failure
+     */
+    virtual int32_t deinit() {
```
style as in the above file `{` on the new line for body functions

0xc0170 commented:
@mazimkhan @hanno-arm you might be interested in this addition, please review if that is the case

ciarmcom commented:
ARM Internal Ref: IOTSSL-1938

deepikabhavnani commented on features/crc/soft/HalfByteCRC.h:
```diff
@@ -0,0 +1,149 @@
+ */
+
+#ifndef MBED_HALF_BYTE_CRC_H
+#define MBED_HALF_BYTE_CRC_H
```
@geky - Please check if this implementation will suffice the CRC32 requirement for littlefs.
Tested with: https://gist.github.com/deepikabhavnani/ff6fc74c4150cff36d398c1d2a4e7f59

geky replied:
Looks good to me! Only thought is that the name `HalfByteCRC` is a bit confusing. What about `MediumCRC`. Another thought, what if we renamed `SlowCRC` to `SmallCRC`. Just tossing out ideas, it's up to you.

other terribles names:
- `ModestCRC`
- `PracticalCRC`
- `KindaFastKindaSlowCRC`
- `BestCRC`

geky replied:
Actually, thinking more about it. Maybe we should just leave out `HalfByteCRC` and force LittleFS to use `FastCRC`. It's probably better to have fewer options so there's more chances for the implementation to get reused.

We can also provide a config option for LittleFS to use `SlowCRC` (`SmallCRC`?) in the cases the size penalty is too much.

Thoughts?

deepikabhavnani replied:
>  name HalfByteCRC is a bit confusing

Naming would be good based on speed/technique.
Since we have so many different variations  we can change from fast/slow to technique based.\
Fast CRC - `lookupTable` or `StandardTable` CRC
Slow CRC - `bitwiseCRC` or `tableless` CRC.
HalfByteCRC - Half-byte/Nibble/16-byte entry table are different options for

deepikabhavnani replied:
> Actually, thinking more about it. Maybe we should just leave out HalfByteCRC and force LittleFS to use FastCRC

We can have default as FastCRC and switch to SlowCRC option in targets.json. Selection will be one-time mostly based on target and not application. Correct?
No harm in providing half-byte option as well, we just need to finalize one default.

sbutcher-arm commented:
Hi @0xc0170,

This isn't security or crypto related, so doesn't _require_ a review from @hanno-arm or @mazimkhan, although I don't object to them reviewing it.

Could you please remove the TLS label? Thanks.

geky commented on features/crc/soft/HalfByteCRC.h:
```diff
@@ -0,0 +1,150 @@
+        MBED_ASSERT(buffer != NULL);
+
+        uint8_t *data = static_cast<uint8_t *>(buffer);
+        uint32_t remainder = p_crc == 0x0 ? _inital_value : ~p_crc;
```
p_crc == 0x0 might be a valid intermediary value for the crc, this would break if that lined up with a `compute_partial`  block. You can expect the user to either initialize it themselves, or call `compute` first

deepikabhavnani replied:
Asking user to set initial value will be better then calling compute first. Will update this.

geky commented on features/crc/soft/HalfByteCRC.h:
```diff
@@ -0,0 +1,150 @@
+        uint8_t *data = static_cast<uint8_t *>(buffer);
+        uint32_t remainder = _inital_value;
+
+        for (crc_data_size_t byte = 0; byte < size; byte++) {
```
Can `compute` just call `compute_partial` now?

geky commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,216 @@
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, uint32_t *crc)
+    {
+        MBED_ASSERT(crc != NULL);
```
Now that these aren't templates, could the implementations be moved back into C++ files?

deepikabhavnani replied:
FastCRC implementation is template based

geky commented on features/crc/soft/HalfByteCRC.h:
```diff
@@ -0,0 +1,150 @@
+            remainder = _crc_table[((remainder ^ (data[byte] >> 4)) & 0x0F)] ^ (remainder >> 4);
+        }
+
+        *crc = (remainder ^ _final_xor);
```
XORing with `_final_xor` will also cause problems in partial computes. Not sure the best way to handle this. You could have `compute_partial_start` and `compute_partial_stop` for the initializing/finalizing. In LittleFS's case it's just handling the initializing/finalizing itself so you could also just get by without the initializing/finalizing for now.

SenRamakri commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,202 @@
+private:
+    uint32_t _polynomial;
+    uint8_t _width;
+    T _crc_table[256];
```
Shouldn't the CRC table be same for same polynomial? In that case, have you thought of sharing the table for same polynomial types across multiple instances. That can help save lot of memory and will avoid computing the table for same polynomial type in init. May be the assumption here is to use same CRC object across the module(For E.g.: fs, block device), in that case should we protect the interfaces for thread-safety?

deepikabhavnani replied:
Having common CRC table across modules will require proper thread-safety, but I am not sure if any application will have this use case. Can however think of it in future

SenRamakri commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,202 @@
+     *  @note CRC calculations are done byte wise.
+     */
+    virtual int32_t compute(void *buffer, data_size_t size, uint32_t *crc) {
+        MBED_ASSERT(crc != NULL);
```
Isn't MBED_ASSERTs removed for non-debug builds? In that case its better to return an error when you get an invalid parameter instead of working with invalid data(works for release builds).

geky replied:
> In that case its better to return an error when you get an invalid parameter

== more code size

In this case we'll hit a hard fault either way, which is probably clear enough as to what went wrong IMO.

SenRamakri replied:
As of now hitting Hard Fault will put the application into an unrecoverable state, where as if we return an error the application can handle that more gracefully, like logging/reporting the error to the user. And hitting hardfault is not a scenario easily debuggable in a customer/deployed scenario and without a debugger. But I do agree that returning error code will slightly increase the code size.

geky replied:
But this case is a usage error. The correct way to handle this error is to not pass in NULL in the first place.

Point taken for making it easier to debug in the customer scenario.

But I have a hard time imagining the case where someone propagates error codes correctly, but passes NULL into this function.

---

Just food for thought, I don't think this conversation should block this pr

deepikabhavnani commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,145 @@
+     *
+     *  @param  crc  Intial CRC value set by the API
+     */
+    virtual void compute_partial_start(uint32_t *crc)
```
> Maybe it would be better to mark this function as pure virtual so implementers doesn't skip it.

Implementation of partial CRC computation shouldn;t be forced. I am not sure if all hardware CRC implementations will support that.

pan- replied:
From a user and implementer perspective, this is confusing: if *A* and *B* inherits from CRC, when *A* is used as an instance of CRC it may not be replaceable by an instance of *B*.  It is not mentioned that these functions may not be implemented. 

This also has severe implications on usability on hardware CRC if the data used to compute the CRC arrive by chunk: the application should either use a software CRC when a chunk is ready to be processed or recompose the whole data to process before the computation of the CRC if hardware computation is used. 

It may not always be possible to store all the data in memory (as an example consider a large file on disk); the current interface forces the use of Software CRC in such case. 

>  I am not sure if all hardware CRC implementations will support that.

@bulislaw Can we have clarity on this point ?

deepikabhavnani commented:
> Would it be possible to use pre initialized look up tables for FastCRC and HalfByteCRC (in addition of existing methods) ?

Pre-initialized to one default, or pre-initialize all supported and select based on compiler defines? At present I see 3 use cases of CRC (SD-driver uses 7-bit and 16-bit CRC and littlefs uses CRC32), should we have table for all of them? 

If we really need tables, we can have a separate header file just for tables. But we need to decide which polynomial table we should add.

deepikabhavnani commented:
> In the parent class, the CRC is always passed as an uint32_t 

Parent class is kept generic, so that HW CRC class can be derived from that. I saw implementation of few targets and hence selected it as uint32_t, since all lower level implementation will be in C.
What do you suggest in this case, should the base class be template as well or change derived class? Do we have any idea of HW CRC class (which might be part of HAL)?

pan- commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,145 @@
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC value (intermediate CRC)
```
Could you specify that `crc` may not be a valid CRC and therefore shouldn't be used as one ?

pan- commented:
> What do you suggest in this case, should the base class be template as well or change derived class?

I'd suggest to use `crc_polynomial_t` as the template parameter of `FastCRC` and `SlowCRC` rather than the receiving CRC type. It may also be interesting to have a function in the CRC interface that query the CRC width. Type erasing the crc parameter - used within the various CRC functions - with a `void*` can be considered too if we want a single interface that handles everything.

deepikabhavnani commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,145 @@
+/** @{*/
+#include <stdint.h>
+
+typedef uint64_t crc_data_size_t;
```
You mean add this define to every derived class?

pan- replied:
In that context derived classes would have access to the typedef; no reason to define it in every class. Inside CRC is enough.

pan- replied:
In that context derived classes would have access to the typedef; no reason to define it in every class. Inside CRC is enough.

deepikabhavnani commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,145 @@
+     *  @param  crc  CRC
+     *  @return  0 on success, negative error code on failure
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, uint32_t *crc) = 0;
```
Decision for this API will depend on compute_partial_xxx functions. If those are made pure virtual and implemented by all derived classes then we have have default implementation for this.

From my understanding CRC class must provide compute function, hence I kept it as pure virtual Though software CRC implementation uses compute_partial, but that might not be the case with HW CRC as well. We may have cases where continuous stream of data is transferred and we need to compute CRC in parts, but many applications also have small CRC where partial computation is not required.

bulislaw commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,223 @@
+        crc_polynomial_type_t polynomial = this->get_polynomial_type();
+
+        if (CRC_7BIT_SD == polynomial) {
+            return 0x9;             // x7+x3+1;
```
I would move the polynomials expansion to the type enum.

bulislaw commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,223 @@
+     *  @param [OUT] crc  CRC
+     *  @return  0 on success, negative error code on failure
+     */
+    virtual int32_t compute(void *buffer, crc_data_size_t size, crc_size_t *crc) = 0;
```
Why are we using/calling the `crc_size_t`? It reassembles library type `ssize_t` and I would expect it to represent size. How about `crc_sum_t`?

deepikabhavnani replied:
My preference would be to use uint32_t, but yes can have it as crc_sum_t

bulislaw commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,223 @@
+        return (width <= 8 ? 1 : (width <= 16 ? 2 : 4));
+    }
+
+    /** Get the top bit of current CRC
```
I don't get that, what's the use case?

deepikabhavnani replied:
Top bit will be used in software computation of CRC calculation. While performing the division operation we do not shift one bit at a time, the divisor is aligned to the first 1 of dividend. Here we perform right shifts and align dividend first 1 at the top-bit to perform division.

bulislaw replied:
That seems like we are exposing internal functions as API. Also 'top bit' is msb? maybe we should stick to msb/lsb?

deepikabhavnani replied:
Top bit is not necessary MSB, in case of 6-bit CRC it will be 6th bit on unsigned char and not MSB bit.

deepikabhavnani replied:
> That seems like we are exposing internal functions as API.

Yes, neither did I wanted them to be APIs (were implemented as macros earlier). But was suggested to have them as functions. I have two options 
1. Protected (as used by all derived classes in software implementation)
2. Move to software implementation class (multiple copies)

bulislaw replied:
Protected sounds good.

bulislaw commented on features/crc/mbed_crc.h:
```diff
@@ -0,0 +1,223 @@
+        return (width < 8 ? (1u << 7) : (1u << (width - 1)));
+    }
+
+    /** Get the CRC data mask
```
I'm not sure why that's necessary?

deepikabhavnani replied:
CRC result is 32-bit, but polynomial is not always 32-bit. We need to higher mask the bits to get correct CRC in result.

bulislaw replied:
That shouldn't be part of public API then?

deepikabhavnani replied:
These API's will be used by derived class, if not private then
1. Protected.
2. Move them to all three software implementation class.

bulislaw replied:
Protected sounds good.

bulislaw commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,267 @@
+
+/** Templated FastCRC Class
+  *
+  *  @note typename must be consistent with polynomial length.
```
Yeah, I would assume it should.

bulislaw commented on features/crc/soft/FastCRC.h:
```diff
@@ -0,0 +1,244 @@
+
+#include "mbed_crc.h"
+
+/** Templated FastCRC Class
```
Can we have description of what FastCRC is here?

bulislaw commented on features/crc/soft/SlowCRC.h:
```diff
@@ -0,0 +1,214 @@
+
+#include "mbed_crc.h"
+
+/** SlowCRC Class
```
Can we have description of what SlowCRC is here?

bulislaw commented:
> One thing I would like to change is to simplify the APIs. Especially the CRC class selection, now we have 3 different CRC backends (fast, slow, halfbyte) and we will also add HW. I don't think we should ask user to figure out what they need. Can we provide a default option? I would like to have a central point (facade?) that would provide API and delegate the computation either to default or to HW if supported for given target and requested CRC combination.

@SenRamakri what do you think?

deepikabhavnani commented:
Converted software CRC classes as non-template classes (user can easily switch) and can have default class.
Moved files to driver folder, and rebased on top of master.

@bulislaw - Same header filename conflict still exists, hence cannot have crc.h as class header file name. CRC.h exists in multiple target folders (not just K64F), hence not updating them.

deepikabhavnani commented:
One more round of review pls @geky @bulislaw @pan- @sg- @0xc0170

geky commented on drivers/MbedCRC.h:
```diff
@@ -0,0 +1,29 @@
+
+#if DEVICE_CRC
+#include "HwCRC.h"
+typedef mbed::BaseCRC CRC;
```
Should this be `typedef mbed::HwCRC CRC;`?

bulislaw commented on drivers/SlowCRC.h:
```diff
@@ -0,0 +1,109 @@
+/* mbed Microcontroller Library
```
Why do we have that if it's 'slow'? Does it have any advantages on 'fast'?

deepikabhavnani replied:
Slow CRC does not create lookup table (less of RAM/ROM space), hence can be used by devices with limited memory.

bulislaw replied:
Can we call it small crc? And document these differences somewhere? I would say we could add it to file doxy and handbook later when we have a page.

deepikabhavnani replied:
Small CRC will be misleading (related to length of CRC polynomial). I can rename it to BitwiseCRC, if that is fine.

deepikabhavnani replied:
Yes I will add it to handbook (with example), once this PR is ready to merge.

deepikabhavnani commented on drivers/MbedCRC.h:
```diff
@@ -0,0 +1,29 @@
+
+#include <stdint.h>
+
+#if DEVICE_CRC
```
I would like to leave it default this way, and create class later on (based on HW), as it might result in lot more changes later (specially in constructor) if we have different polynomials supported by different HW.

bulislaw replied:
That's also fine.

0xc0170 replied:
Not sure why we would leave this condition here. The class does not exist, neither device_has CRC. Once it is defined, this would be added here. Otherwise it is dead code, if I define CRC in `device_has` or via command line, it does not build.

deepikabhavnani replied:
Agree @0xc0170  Entire device CRC code below is dead code, but it was kept to add default CRC as HW (which is not present now) and switch back to FastCRC as default.
`+#if DEVICE_CRC
 +#include "HwCRC.h"
 +typedef mbed::HwCRC RawCRC;`

I can either added error here (#error HwCRC not supported), or leave this code commented.

bulislaw commented:
We should add `MbedCRC.h` to `mbed.h`.

0xc0170 commented on drivers/FastCRC.cpp:
```diff
@@ -0,0 +1,184 @@
+    }
+}
+
+FastCRC::~FastCRC() {
```
`{` new line

0xc0170 commented on drivers/FastCRC.cpp:
```diff
@@ -0,0 +1,184 @@
+    deinit();
+}
+
+int32_t FastCRC::init(void)
```
What is the intention with init/deinit methods, instead of doing this in ctor/dtor (dtor does call deinit, but init is postponed).

deepikabhavnani replied:
Yes init is not mandatory for all techniques, we create table during init and allocate memory, which may result into error.
Also, some hardware CRCs require special init.

0xc0170 commented on drivers/BaseCRC.h:
```diff
@@ -0,0 +1,238 @@
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC value is intermediate CRC value filled by API.
+     *  @return  0  on success or a negative error code on failure
```
Shall we define negative error codes, or -1 does it all?

deepikabhavnani replied:
-1 is the only error code in Software CRC implementation. However, I can use "Not supported" as additional error code, if we have common error codes, but nothing special at present.

deepikabhavnani commented:
@bulislaw - Please approve if all your issues are addressed. Thanks

0xc0170 commented:
/morph build

cmonr commented:
@deepikabhavnani Will 5911 supersede this PR?

deepikabhavnani commented:
@cmonr - yes it will. We can close this once 5911 is approved.

deepikabhavnani commented:
Closing this, as PR #5911 will be merged after approvals
